forge-ai/src/main/java/forge/ai/AiController.java

```diff
@@ -1565,79 +1596,94 @@ private SpellAbility chooseSpellAbilityToPlayFromList(final List<SpellAbility> a
-                    } else if (CardLists.filter(player.getZone(ZoneType.Graveyard).getCards(), CardPredicates.Presets.CREATURES).size() > 4) {
-                        if (player.getCreaturesInPlay().size() >= 4) // it's good minimum
-                            continue;
-                        else if (!sa.getHostCard().isPermanent() && sa.canCastTiming(player) && ComputerUtilCost.canPayCost(sa, player, sa.isTrigger()))
-                            aiPlayDecision = AiPlayDecision.WillPlay;// needs tuneup for bad matchups like reanimator and other things to check on opponent graveyard
-                    } else {
-                        continue;
-                    }
                 }
-            }
 
-            sa.setActivatingPlayer(player, true);
-            SpellAbility root = sa.getRootAbility();
+                sa.setActivatingPlayer(player, true);
+                SpellAbility root = sa.getRootAbility();
 
-            if (root.isSpell() || root.isTrigger() || root.isReplacementAbility()) {
-                sa.setLastStateBattlefield(game.getLastStateBattlefield());
-                sa.setLastStateGraveyard(game.getLastStateGraveyard());
-            }
-            //override decision for living end player
-            AiPlayDecision opinion = useLivingEnd && AiPlayDecision.WillPlay.equals(aiPlayDecision) ? aiPlayDecision : canPlayAndPayFor(sa);
+                if (root.isSpell() || root.isTrigger() || root.isReplacementAbility()) {
+                    sa.setLastStateBattlefield(game.getLastStateBattlefield());
+                    sa.setLastStateGraveyard(game.getLastStateGraveyard());
+                }
+                //override decision for living end player
+                AiPlayDecision opinion = useLivingEnd && AiPlayDecision.WillPlay.equals(aiPlayDecision) ? aiPlayDecision : canPlayAndPayFor(sa);
 
-            // reset LastStateBattlefield
-            sa.clearLastState();
-            // PhaseHandler ph = game.getPhaseHandler();
-            // System.out.printf("Ai thinks '%s' of %s -> %s @ %s %s >>> \n", opinion, sa.getHostCard(), sa, Lang.getPossesive(ph.getPlayerTurn().getName()), ph.getPhase());
+                // reset LastStateBattlefield
+                sa.clearLastState();
+                // PhaseHandler ph = game.getPhaseHandler();
+                // System.out.printf("Ai thinks '%s' of %s -> %s @ %s %s >>> \n", opinion, sa.getHostCard(), sa, Lang.getPossesive(ph.getPlayerTurn().getName()), ph.getPhase());
 
-            if (opinion != AiPlayDecision.WillPlay)
-                continue;
+                if (opinion != AiPlayDecision.WillPlay)
+                    return 0;
 
-            return sa;
+                spells.add(sa);
+                return 0;
+            }));
+        }
+        //timeout 5 seconds? even the AI don't acquire all, there should be SA to cast if valid
+        CompletableFuture<?>[] futuresArray = futures.toArray(new CompletableFuture<?>[0]);
```

Comment:
please make it an opt-in option or users will just complain about passive AI in some cases and we can't determine easily if API logic was bad or just boardstate too messy

personally I prefer to wait instead of taking a suboptimal play anyway

Reply:
I can add that option 5 sec, 10 sec, maybe for off 1 hour is good enough for you to timeout?

Reply:
1 hour probably means deadlock haha
5sec, 1min, off or something?

Reply:
I'll just put 1 hour for off (most likely it will be finish unless an error occurred)